### PR TITLE
LED blinking with timer (steady blinking + blinking in stand-by mode)

### PR DESCRIPTION
**Pull request by EmilieNumworks**
Merge PR #314, thank you @boricj !

### CODE REVIEW

File: ion/include/ion/led.h [hunk 1]
```diff
@@ -6,7 +6,9 @@
 namespace Ion {
 namespace LED {
 
+KDColor getColor();
 void setColor(KDColor c);
+void setBlinking(float period, float dutyCycle);
```
Comment (Ecco) [inline ion/include/ion/led.h]: `float period` switch to an integer, such as uint16_t periodInMiliseconds

File: ion/src/device/device.cpp [hunk 2]
```diff
@@ -279,15 +281,33 @@ void initClocks() {
...
   // APB2 bus
   RCC.APB2ENR()->set(0x00008000); // Reset value
 
-  // AHB1
-  RCC.APB1ENR()->set(0x00000400);
-
-  // AHB1 bus
-  RCC.AHB1ENR()->set(0); // Reset value
+  if (keepLEDAwake) {
+    /* TODO: enter sleep mode even if the LED is used and enable LED in low
```
Comment (Ecco) [inline ion/src/device/device.cpp]: Discard these comments, that's just not true 😄

File: ion/src/device/device.cpp [hunk 3]
```diff
@@ -279,15 +281,33 @@ void initClocks() {
...
+     * power mode. */
+    // Keep the LED going: peripheral clock enable in low power mode register
+    /*RCC.APB1LPENR()->setTIM3LPEN(true);
+    RCC.AHB1LPENR()->setGPIOBLPEN(true);
+    RCC.AHB1LPENR()->setGPIOCLPEN(true);*/
+
+    // APB1
+    class RCC::APB1ENR apb1enr(0x00000400);
```
Comment (Ecco) [inline ion/src/device/device.cpp]: Factorize code.
```cpp
class RCC::APB1ENR apb1enr(0x00000400);
if () {
 apb1enr.setFooBar()
}
RCC.APB1ENR()->set(apb1enr);
```

File: ion/src/device/device.h [hunk 4]
```diff
@@ -4,6 +4,8 @@
 namespace Ion {
 namespace Device {
 
+constexpr int SYSBUS_FREQ = 96*1000*1000;
```
Comment (Ecco) [inline ion/src/device/device.h]: Coding Style. Not sure this needs to be in the header by the way.

File: ion/src/device/led.cpp [hunk 5]
```diff
@@ -1,16 +1,55 @@
...
+   * Consequently, we do not use PWM to display the right color anymore.
+   * Instead we 'project the color on 3 bits' : RED LED is active or not etc. */
+  Ion::LED::Device::setColorStatus(Ion::LED::Device::Color::RED, sLedColor.red() > 0);
+  Ion::LED::Device::setColorStatus(Ion::LED::Device::Color::GREEN, sLedColor.green() > 0);
+  Ion::LED::Device::setColorStatus(Ion::LED::Device::Color::BLUE, sLedColor.blue() > 0);
+
+  Device::setPeriodAndDutyCycles(Device::Mode::BLINK, dutyCycle, dutyCycle, dutyCycle, period);
+}
+
+/*void Ion::LED::setCharging(bool isPluggedIn, bool isCharging) {
```
Comment (Ecco) [inline ion/src/device/led.cpp]: Remove

File: ion/src/device/led.cpp [hunk 6]
```diff
@@ -81,12 +103,53 @@ void initTimer() {
...
+    case Ion::LED::Device::Color::GREEN:
+      TIM3.CCMR()->setOC4M(enable ? TIM<Register16>::CCMR::OCM::PWM1 : TIM<Register16>::CCMR::OCM::ForceInactive);
+      break;
+    case Ion::LED::Device::Color::BLUE:
+      TIM3.CCMR()->setOC3M(enable ? TIM<Register16>::CCMR::OCM::PWM1 : TIM<Register16>::CCMR::OCM::ForceInactive);
+      break;
+  }
+}
+
+/*void enforceState(bool red, bool green, bool blue) {
```
Comment (Ecco) [inline ion/src/device/led.cpp]: Discard

File: ion/src/device/led.h [hunk 7]
```diff
@@ -14,10 +14,23 @@ namespace Device {
  *  PC7 | LED red           | Alternate Function 2  | TIM3_CH2
  */
 
+enum Mode {
```
Comment (Ecco) [inline ion/src/device/led.h]: Use enum class

File: ion/src/device/led.h [hunk 8]
```diff
@@ -14,10 +14,23 @@ namespace Device {
...
+  PWM,
+  BLINK
```
Comment (Ecco) [inline ion/src/device/led.h]: Coding style, `Blink`

File: ion/src/device/led.h [hunk 9]
```diff
@@ -14,10 +14,23 @@ namespace Device {
...
+};
+
+enum Color {
```
Comment (Ecco) [inline ion/src/device/led.h]: Idem

File: ion/src/device/led.h [hunk 10]
```diff
@@ -14,10 +14,23 @@ namespace Device {
...
+  RED,
+  GREEN,
+  BLUE
+};
+
 void init();
 void shutdown();
+void setPeriodAndDutyCycles(Mode mode, float dutyCycleRed, float dutyCycleGreen, float dutyCycleBlue, float period = 0.0f);
+void setColorStatus(Color color, bool enable);
```
Comment (Ecco) [inline ion/src/device/led.h]: That name really doesn't help. What does this function do?

File: ion/src/device/led.cpp [hunk 11]
```diff
@@ -81,12 +103,53 @@ void initTimer() {
...
+      period *= TIM3_FREQ; // as we pre-scaled, we update the period
+      break;
+  }
+
+  TIM3.CCR2()->set(dutyCycleRed*period);
+  TIM3.CCR3()->set(dutyCycleBlue*period);
+  TIM3.CCR4()->set(dutyCycleGreen*period);
+}
+
+void setColorStatus(Color color, bool enable) {
```
Comment (Ecco) [inline ion/src/device/led.cpp]: I quite don't like this function. It yields a function call and a switch when previously we only required a simple write at a given address.

File: ion/src/device/led.cpp [hunk 12]
```diff
@@ -81,12 +103,53 @@ void initTimer() {
...
+  setColorStatus(Ion::LED::Device::Color::BLUE, false);
 }
 
-void enforceState(bool red, bool green, bool blue) {
+/* Pulse width modulation mode allows you to generate a signal with a
+ * frequency determined by the value of the TIMx_ARR register and a duty cycle
+ * determined by the value of the TIMx_CCRx register. */
+
+void setPeriodAndDutyCycles(Mode mode, float dutyCycleRed, float dutyCycleGreen, float dutyCycleBlue, float period) {
+  constexpr int TIM3_FREQ = 4*1000;
```
Comment (Ecco) [inline ion/src/device/led.cpp]: Coding Style.

**Comment by matheod** [comment]
J'ai très peu de recul donc je signale juste au cas où. Si je ne dis pas de bétise, la feature pour afficher l'état de chargement de la batterie via la led a été merge, sauf que je viens de tester et aucune led ne s'affiche. Je me demande si le travail autour du mode examen n'aurait pas cassé la feature sur l'affichage du chargement. Mais je dis peut être n'importe quoi :)

**Comment by Ecco** [comment]
Hi @matheod , We're welcoming contributors here who don't speak french so we'd appreciate it if we could stick to english. Thanks! That being said:

> the feature to display the state of loading of the battery via the led was merged

No it has not 😄

**Pull request merged**